js/ad_manager_vast.js

```diff
@@ -1616,6 +1645,218 @@ OO.Ads.manager(function(_, $) {
+    var _findVMAPTrackingEvents = _.bind(function(adBreakElement) {
+      var trackingEventsElement = $(adBreakElement).find("vmap\\:TrackingEvents, TrackingEvents");
+      var VMAPTrackingEventsElement = _.filter(trackingEventsElement.toArray(), function(trackingEventElement) {
+        return (trackingEventElement.tagName.toLowerCase().indexOf("vmap:") > -1);
+      }, this);
+      return VMAPTrackingEventsElement;
+    }, this);
+
+    /**
+     * Helper function to convert VMAP tracking events into objects with attributes as properties.
+     * @private
+     * @method Vast#_parseVMAPTrackingEvents
+     * @param {object} trackingEventsElement The tracking events element
+     * @returns {object[]} The array of tracking event objects.
+     */
+    var _parseVMAPTrackingEvents = _.bind(function(trackingEventsElement) {
+      var trackingEvents = [];
+      var trackingElements = $(trackingEventsElement).find("vmap\\:Tracking, Tracking");
+      if (trackingElements.length > 0) {
+        _.each(trackingElements, function(trackingElement) {
+          var tracking = {};
+          trackingEvents.push(tracking);
+          tracking.url = $(trackingElement).text();
+          tracking.eventName = $(trackingElement).attr("event");
+        }, this);
+      }
+      return trackingEvents;
+    }, this);
+
+    /**
+     * Convert the adBreak attributes into an ad object that will be passed into _onVastResponse().
+     * @private
+     * @method Vast#_convertToAdObject
+     * @param {object} adBreak The adBreak object
+     * @returns {object} null if the timeOffset attribute does not match any format. Otherwise, the
+     * ad object is returned.
+     */
+    var _convertToAdObject = _.bind(function(adBreak) {
+      var adObject = {
+        /*
+         *ad_set_code: "",
+         *click_url: "",
+         *expires: 0,
+         *first_shown: 0,
+         *frequency: 2,
+         *public_id: "",
+         *signature: "",
+         *tracking_url: [],
+         *type: "",
+         *url: ""
+         */
+        time: 0,
+        position_type: "t",
+      };
+      if (adBreak || adBreak.timeOffset) {
+        switch(true) {
+          // case: "start"
+          case /^start$/.test(adBreak.timeOffset):
+            adObject.position_type = "t";
+            adObject.time = 0;
+            break;
+          // case: "end"
+          case /^end$/.test(adBreak.timeOffset):
+            adObject.position_type = "t";
+            adObject.time = (this.amc.movieDuration + 1) * 1000;
+            break;
+          // case: hh:mm:ss.mmm | hh:mm:ss
+          case /^\d{2}:\d{2}:\d{2}\.000$|^\d{2}:\d{2}:\d{2}$/.test(adBreak.timeOffset):
+            adObject.position_type = "t";
+            adObject.time = _convertTimeStampToSeconds(adBreak.timeOffset);
+            break;
+          // case: [0, 100]%
+          case /^\d{1,3}%$/.test(adBreak.timeOffset):
+            // TODO: test percentage > 100
+            adObject.position_type = "t";
```

Comment:
So according to our docs (https://wiki.corp.ooyala.com/display/idocs/Player+V4+IMA+Review+Draft#link-talk-1417), position_type "p" for percentages should also be acceptable, but when I tried testing it with "p" and passing the appropriate percentage, it didn't seem to work, but yeah we can pull it out of the switch if only is "t" is available. I'll double check